### ISSUE TITLE
Demystifying Convolutional Neural Networks

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8627]
Demystifying Convolutional Neural Networks<br>
<br>
https://ift.tt/npNDPeC<br>
<br>
Aegeus Zerium<br>
<br>
<div>
<div>
<div>
<div>
<div>
<h1>Demystifying Convolutional Neural Networks</h1>
<p>An Intuitive Explanation of Convolutional Neural Networks.</p>
<h1>Definition:</h1>
<p>Simply put, a <a rel="noopener ugc nofollow" href="https://en.wikipedia.org/wiki/Convolutional_neural_network" target="_blank"><strong>Convolutional Neural Network</strong></a> is a Deep learning model or a multilayered percepteron similar to Artificial Neural Networks which is most commonly applied to analyzing visual imagery. The founding father of Convolutional Neural Networks is the well known computer scientist working in Facebook <a rel="noopener ugc nofollow" href="https://en.wikipedia.org/wiki/Yann_LeCun" target="_blank"><strong>Yann LeCun</strong></a> who was the first one to use them to solve the hand written digits problem using the famous <a rel="noopener ugc nofollow" href="http://yann.lecun.com/exdb/mnist/" target="_blank"><strong>MNIST</strong></a> Dataset.</p>
<figure>
<figcaption>Yann LeCunn</figcaption>
</figure>
<p>Convolutional Neural Networks were inspired by biological processes in that the connectivity pattern between neurons resembles the organization of the animal visual cortex.</p>
<figure>
<figcaption>Visual structure of the human brain</figcaption>
</figure>
<p>Individual cortical neurons respond to stimuli only in a restricted region of the visual field known as the receptive field. The receptive fields of different neurons partially overlap such that they cover the entire visual field.</p>
<figure>
<figcaption>Computer vision vs Human vision</figcaption>
</figure>
<p>As you can see we can&rsquo;t possibly talk about any type of Neural Networks without mentioning a little bit of neuroscience and how the human body (especially the brain) and its functions have been the primary inspiration for the creation of various Deep learning models.</p>
<h1>The architecture of ConvNets:</h1>
<figure>
<figcaption>ConvNet architecture</figcaption>
</figure>
<p>As you can see in the illustration above a ConvNet architecture is very similar to the regular ANN architecture especially in the last layers of the network namely the Fully connected layers area, you will also notice that a ConvNet accepts a volume as an input instead of a vector.</p>
<p>let&rsquo;s now explore the layers that constitute a ConvNet and the mathematical operations that the latter goes through to visualize and classify pictures based on the features and attributes it has learnt during the training process.</p>
<h1>Input Layer:</h1>
<p>The input layer is mostly an n &times; m &times; 3 RGB (short for Red, Green and Blue) image(s) unlike an Artificial Neural Networks which gets fed with a n &times; 1 vector, nothing hard to grasp here.</p>
<figure>
<figcaption>RGB image</figcaption>
</figure>
<h1>Convolution Layer:</h1>
<p>In the <strong>Convolution layer</strong> we compute the output of the dot product between an area of the input image(s) and a weight matrice called a <strong>filter</strong>, the filter will slide through out the whole image repeating the same dot product operation. Two things that should be mentioned:</p>
<ul>
<li>The filter must have the same number of channels as the input image.</li>
<li>it&rsquo;s commonly known that the deeper you go into the Network the more filters you use the intuition behind it is that the more filters we have the more edge and feature detection you&rsquo;ll get</li>
</ul>
<figure>
<figcaption>Forward Convolution Operation</figcaption>
</figure>
<p>We calculate the dimensions of the output of the Convolution layer:</p>
<p><strong><em>Output Width:</em></strong></p>
<p><strong><em>Output Height:</em></strong></p>
<p>where:</p>
<ul>
<li>
<em>W</em> : the width of the input image</li>
<li>
<em>H</em> : the height of the input image</li>
<li>
<em>Fw</em> : the width of the filter or kernel</li>
<li>
<em>Fh</em> : the height of the filter</li>
<li>
<em>P</em> : padding</li>
<li>
<em>S</em> : stride</li>
</ul>
<p>The number of channels of the Convolution layer output equals to the <strong>number of filters</strong> used during the convolution operation.</p>
<h2>Why Convolutions ?</h2>
<p>You are probably asking yourself, why do we use convolutions in the first place ? why not flatten the input images from the beginning ? well if we do that we will end up with a massive number of parameters that need to be trained and most of us don&rsquo;t have the computational power that will solve our computationally expensive task in the fastest way possible. In addition, with the fewer parameters that the ConvNets have we can avoid overfitting.</p>
<h1>Pooling Layer:</h1>
<p>There are two widely used types of pooling, average pooling and max pooling where the latter being the most used of the two. The pooling layer is used to reduce the spatial dimensions, but not depth, on a convolutional neural network. When using the max pooling layer we take the highest number (the most responsive area in the image) of the input&rsquo;s area (an n &times; m matrice), whereas when we use the average pooling layer we take the mean of the input area instead.</p>
<figure>
<figcaption>Max Pooling</figcaption>
</figure>
<h2>Why Pooling ?</h2>
<p>One of the core goals of the pooling layer (max pooling in this case) is to provide spatial variance, which simply means that you or the machine will be capable of recognizing an object as an object even when its appearance varies in some way. for more in depth explanation of the pooling layer check this rigourous <a rel="noopener ugc nofollow" href="http://yann.lecun.org/exdb/publis/pdf/boureau-icml-10.pdf" target="_blank"><strong>paper</strong></a> by Yann LeCunn.</p>
<h1>Non-linearity Layer:</h1>
<p>In the Non-linearity layer we use the ReLU activation function most if not all the time instead of the Sigmoid or Tan-H activation function. The ReLU activation function returns 0 for every negative value in the input image while it returns the same value for every positive value in the input image (for more in depth explanation of activation functions please check this <a rel="noopener ugc nofollow" href="https://blog.goodaudience.com/artificial-neural-networks-explained-436fcf36e75" target="_blank"><strong>article</strong></a> of mine).</p>
<figure>
<figcaption>ReLU activation function</figcaption>
</figure>
<h1>Fully Connected Layer:</h1>
<p>In the FC layer we flatten the output of the last Convolution layer and connect every node of the current layer with the other node of the next layer, Fully connected layer is just another word for the regular Artificial neural network as you will see in the image below. The operations in the fully connected layer are exactly the same as in any artificial neural network:</p>
<figure>
<figcaption>Flattening of the Convolution layer</figcaption>
</figure>
<figure>
<figcaption>Fully connected layer</figcaption>
</figure>
<p>The layers and operations discussed above are the core components of every Convolutional neural network.</p>
<p>Now that we&rsquo;ve discussed the operations that a ConvNet goes through in a forward pass let&rsquo;s jump to the operations that a ConvNet goes through in a backward pass.</p>
<h1>Backpropagation:</h1>
<h1>Fully Connected Layer:</h1>
<p>in the Fully connected layer backpropagation works exactly the same as in any regular artificial neural network, in backpropagation (using gradient descent as an optimization algorithm) we use partial derivatives namely the derivative of the loss function with regard to the weights, in order to calculate the latter we use a well known operation in calculus called <strong>The Chain rule</strong> where we multiply (in the backpropagation context) the derivative of the loss function w.r.t the activated output with the derivative of the activated output w.r.t the non-activated output with the derivative of the non-activated output w.r.t to the weights.</p>
<p>I was going to write in front of each partial derivative its mathematical notation but unfortunately Medium does not natively support writing mathematical expressions which is quite infuriating to be honest, anyway here is how we can summarize what we&rsquo;ve just said in plaintext above with this mathematical expression below:</p>
<figure>
<figcaption>Backpropagation illustration</figcaption>
</figure>
<p>after calculating the gradient we substract it from the initial weights to get newly optimized ones:</p>
<p>where:</p>
<ul>
<li>
<em>&theta;i+1</em> : optimized weights</li>
<li>
<em>&theta;i</em> : initial weights</li>
<li>
<em>&alpha;</em> : learning rate</li>
<li>
<em>&nabla;J(&theta;i)</em> : gradient of the loss function</li>
</ul>
<figure>
<figcaption>Gradient descent</figcaption>
</figure>
<p>In the animation below, gradient descent is applied to linear regression, you can clearly see that the more the cost function gets minimized the better the linear model fits the data.</p>
<figure>
<figcaption>Gradient descent applied to linear regression</figcaption>
</figure>
<p>note that you should be careful with choosing the value of the learning rate, a very high learning rate could cause the gradient to overshoot the target minimum.</p>
<figure>
<figcaption>Small learning rate vs Big learning rate</figcaption>
</figure>
<p>In all optimization tasks ,whether in physics, economics or Computer science, partial derivatives are overwhelmingly used, partial derivatives are primarily used to calculate the rate of change of a dependent variable <em>f(x,y,z)</em> with regard to one of its independent variables while the rest of the variables remain constant. for example imagine you own a share of a company, the stocks of the latter will go up or down based on multiple factors (security, politics, sales revenue etc &hellip;), to implement partial derivatives on your situation you would calculate how much the stock price of your company change if security (for example) get affected while others factors remain constant and repeat the same process with each and every other factor.</p>
<h1>Pooling Layer:</h1>
<p>In the Max Pooling layer the gradient gets backpropagated through the maximum values only since changing them slightly won&rsquo;t affect the output. In the process we replace the maximum values before max pooling with 1 and set all the non maximum values to zero then use the <strong>Chain rule</strong> to multiply the gradient by them.</p>
<figure>
<figcaption>Backpropagation through the Pooling layer</figcaption>
</figure>
<p>Unlike the max pooling layer, in the average pooling layer the gradient passes through all the inputs (before average pooling) the maximum and the non maximum ones.</p>
<h1>Convolution layer:</h1>
<p>You are probably asking yourself right now, if the forward pass of a convolution layer is a convolution then what is its backward pass ? luckily, its backward pass is also a convolution (as you can clearly see below) so you don&rsquo;t need to worry about learning new set of hard to grasp mathematical operations.</p>
<figure>
<figcaption>Backpropagation through the Convolution layer</figcaption>
</figure>
<p>where:</p>
<ul>
<li>
<em>&part;hij</em>: the derivative of the loss function w.r.t the output of the convolution layer</li>
</ul>
<p>This is in a nutshell how backpropagation works in a Convolution layer.</p>
<p>Now that you have a robust theoretical understanding of Convolutional Neural Networks let&rsquo;s build our first ConvNet with TensorFlow.</p>
<h1>Convolutional Neural Network with TensorFlow:</h1>
<h1>What is Tensorflow ?</h1>
<p><a rel="noopener ugc nofollow" href="https://opensource.com/article/17/11/intro-tensorflow" target="_blank"><strong>TensorFlow</strong></a> is an open source software library for numerical computation using data-flow graphs. It was originally developed by the Google Brain Team within Google&rsquo;s Machine Intelligence research organization for machine learning and deep neural networks research.</p>
<h1>What is a Tensor ?</h1>
<p>A <a rel="noopener ugc nofollow" href="https://en.wikipedia.org/wiki/Tensor" target="_blank"><strong>tensor</strong></a> is an organized multidimensional array of numerical values. The order (also degree or rank) of a tensor is the dimensionality of the array needed to represent it.</p>
<figure>
<figcaption>Types of tensors</figcaption>
</figure>
<h1>What is a Computational Graph ?</h1>
<p><a rel="noopener ugc nofollow" href="http://www.cs.columbia.edu/~mcollins/ff2.pdf" target="_blank"><strong>Computational graphs</strong></a> are a powerful formalism that have been extremely fruitful in deriving algorithms and software packages for neural networks and other models in machine learning. The basic idea in a computational graph is to express some model &mdash; for example a feedforward neural network &mdash; as a directed graph expressing a sequence of computational steps. Each step in the sequence corresponds to a vertex in the computational graph; each step corresponds to a simple operation that takes some inputs and produces some output as a function of its inputs.</p>
<p>In the illustrated graph below we have two inputs <em>w1=x</em> and <em>w2=y</em>, the inputs will flow through the graph where each node in the graph is a mathematical operation to give us the following outputs:</p>
<ul>
<li>
<em>w3 = cos(x)</em> where the operation is the Cosine trigonometric function</li>
<li>
<em>w4 = sin(x)</em> where the operation is the Sine trigonometric function</li>
<li>
<em>w5 = w3&#8729;w4</em> where the operation is multiplication</li>
<li>
<em>w6 = w1/w2</em> where the operation is division</li>
<li>
<em>w7 = w5+w6</em> where the operation is addition</li>
</ul>
<p>Now that we understand what a computational graph is, let&rsquo;s build our own in tensorflow, we will build the same one above.</p>
<h1>Code:</h1>
<h1>Visualization with Tensorboard:</h1>
<h1>What is Tensorboard ?</h1>
<p><a rel="noopener ugc nofollow" href="https://github.com/tensorflow/tensorboard" target="_blank"><strong>TensorBoard</strong></a> is a suite of web applications for inspecting and understanding your TensorFlow runs and graphs, it&rsquo;s one of the biggest edges that Google&rsquo;s TensorFlow has over Facebook&rsquo;s <a rel="noopener ugc nofollow" href="https://pytorch.org/" target="_blank"><strong>Pytorch</strong></a>.</p>
<figure>
<figcaption>The previous code visualized in Tensorboard</figcaption>
</figure>
<p>Now that you have a robust understanding of ConvNets, TensorFlow and TensorBoard, let&rsquo;s build our first ConvNet that will recognize hand written digits using the <strong>MNIST</strong> dataset.</p>
<figure>
<figcaption>MNIST dataset</figcaption>
</figure>
<p>The architecture of our Convnet will be a set of convolution, max-pooling and non linearity operation layers similar to the <a rel="noopener ugc nofollow" href="http://yann.lecun.com/exdb/lenet/" target="_blank"><strong>LeNet-5</strong></a> architecture.</p>
<figure>
<figcaption>Convolutional Neural Network 3D Simulation</figcaption>
</figure>
<h1>Code:</h1>
<p>The code is lengthy but shouldn&rsquo;t be intimidating if you break it piece by piece.</p>
<p>In case you run the program your results should be like this:</p>
<p>We have just finished building our first convolutional neural network, as you can see in the results above, the accuracy has dramatically increased from the first step to the last step, but still there is more room for improvement of our ConvNet.</p>
<p>let&rsquo;s now visualize our ConvNet in Tensorboard:</p>
<figure>
<figcaption>ConvNet Visualization</figcaption>
</figure>
<figure>
<figcaption>Accuracy and Loss evaluation</figcaption>
</figure>
<h1>Conclusion:</h1>
<p>Convolutional Neural Networks are powerful deep learning models that are applied in a wide range of fields such as radiology, the use of ConvNets will only increase as the data gets bigger and the problems become more sophisticated and challenging.</p>
<h1>Note:</h1>
<p>You can find the Jupyter Notebook of this article in:</p>
<h1>References:</h1>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Medium<br>
<br>
August 26, 2022 at 05:41PM
